include/seqan3/search/kmer_index/shape_iterator.hpp

```diff
@@ -0,0 +1,86 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Vinzenz May <vinzenz.may AT fu-berlin.de>
+ * \brief iterator class to calculate k-mer hash values.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/search/kmer_index/shape.hpp>
+#include <seqan3/std/ranges>
+
+namespace seqan3
+{
+
+  /*!\brief iterator class for generating hash values used by kmer-index
+   * \ingroup search
+   * \tparam it_t iterator of type std::ForwardIterator. Underlying Container must model seqan3::Semialphabet.
+   */
+
+template <std::ForwardIterator it_t>
+    //!\cond
+    requires Semialphabet<value_type_t<it_t>>
+    //!\endcond
+class shape_iterator
+{
+private:
+    using alphabet_t = value_type_t<it_t>;
+    size_t hash_value{0};
+    shape s;
+
+public:
+    it_t text_left, text_right;
+
+    shape_iterator(it_t it_start, shape s_) : s(s_),
+                                    text_left(it_start),
+                                    text_right(it_start)
+    {
+        if (s.size()) // delete for better performance? - generates seg fauls if s.size() == 0 !
+        {
+            hash_value += to_rank(*(text_right));
+            for (size_t i{0}; i < std::ranges::size(s)-1; ++i)
+            {
+                std::advance(text_right, 1);
+
+                hash_value *= alphabet_size_v<alphabet_t>;
+                hash_value += to_rank(*(text_right));
+            }
+        }
+
+    };
+
+    bool operator==(const it_t& it)           { return text_right == it; }
+    bool operator==(const shape_iterator& it) { return text_right == it.text_left; }
+
+    bool operator!=(const it_t& it)           { return text_right != it; }
+    bool operator!=(const shape_iterator& it) { return text_right != it.text_left; }
+
+    void operator ++()
+    {
+
+        hash_value -= to_rank(*(text_left)) * pow(alphabet_size_v<alphabet_t>, std::ranges::size(s)-1);
```

Comment:
You should store the `pow(alphabet_size_v<alphabet_t>, std::ranges::size(s)-1)` in a member variable since this is a costly operation